pallet/dex/src/lib.rs

```diff
@@ -0,0 +1,525 @@
+	order_type: OrderType,
+	amount_offered: u128,
+	amout_requested: u128,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::{ValueQuery, *},
+		Blake2_128Concat,
+	};
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig {
+		pub asset_ids: Vec<u32>,
+	}
+
+	impl Default for GenesisConfig {
+		fn default() -> Self {
+			GenesisConfig {
+				asset_ids: Default::default(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			let bounded_token_infoes: BoundedVec<u32, ConstU32<{ u32::MAX }>> = self
+				.asset_ids
+				.clone()
+				.try_into()
+				.expect("genesis asset_ids are more than u32::MAX");
+
+			let mut index = 0;
+			self.asset_ids.iter().for_each(|asset_id| {
+				TokenIndex::<T>::insert(asset_id, index);
+				index += 1;
+			});
+
+			TokenInfoes::<T>::put(bounded_token_infoes);
+		}
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	/// The module configuration trait.
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Expose customizable associated type of asset transfer, lock and unlock
+		type Fungibles: Balanced<Self::AccountId>
+			+ Mutate<Self::AccountId, AssetId = u32, Balance = u128>;
+
+		type PrivilegedOrigin: EnsureOrigin<<Self as frame_system::Config>::RuntimeOrigin>;
+	}
+
+	/************* STORAGE ************ */
+	#[pallet::storage]
+	#[pallet::getter(fn user_token_infoes)]
+	pub type UserTokenInfoes<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId, //address
+		Blake2_128Concat,
+		u32, //asset id
+		TokenInfo,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn token_infoes)]
+	pub type TokenInfoes<T: Config> =
+		StorageValue<_, BoundedVec<u32, ConstU32<{ u32::MAX }>>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn token_index)]
+	pub type TokenIndex<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		u32,
+		u64, //token index
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn next_order_index)]
+	pub(super) type NextOrderIndex<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn orders)]
+	pub type Orders<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		u64, //order index
+		OrderOf<T>,
+		OptionQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn next_pair_order_index)]
+	pub(super) type NextPairOrderIndex<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn pair_orders)]
+	pub type PairOrders<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		(u32, u32),
+		BoundedVec<u64, ConstU32<{ u32::MAX }>>,
```

Comment:
same question related u32

Reply:
(u32, u32) is the asset id pair

Reply:
i meant a vec of u32::MAX

Reply:
As I mention [above][1], it's just a phantom parameter and not preallocated AFAIK.

https://github.com/ggxchain/ggxnode/pull/227/files#r1467692347